File: py/wlearn/pipeline.py
from .errors import ValidationError, NotFittedError, DisposedError
from .bundle import encode_bundle, decode_bundle
from .registry import register, load as registry_load

PIPELINE_TYPE_ID = 'wlearn.pipeline@1'


class Pipeline:
    """Pipeline estimator: chains transformer steps + final estimator.

    Supports fit/predict/score with transformer + estimator chains,
    as well as save/load from WLRN bundles.
    """

    def __init__(self, steps):
        """Create a pipeline from (name, estimator) pairs.

        Args:
            steps: list of (name, estimator) tuples
        """
        if not steps:
            raise ValidationError('Pipeline requires at least one step')
        self._steps = list(steps)
        self._fitted = False
        self._disposed = False

    def fit(self, X, y):
        """Fit the pipeline: transform through intermediates, fit last step.

        For intermediate steps that have a transform method:
        - If fit_transform exists, use it (avoids fitting + transforming separately)
        - Otherwise, call fit() then transform()
        The last step is only fitted (not transformed).
        """
        self._ensure_alive()
        current = X
        for i, (_, est) in enumerate(self._steps[:-1]):
            if hasattr(est, 'fit_transform'):
                current = est.fit_transform(current, y)
            else:
                est.fit(current, y)
                current = est.transform(current)
        # Last step: fit only
        _, last = self._steps[-1]
        last.fit(current, y)
        self._fitted = True
        return self

    def _transform_through(self, X):
        """Transform X through all intermediate steps (not the last)."""
        current = X
        for _, est in self._steps[:-1]:
            current = est.transform(current)
        return current

    def predict(self, X):
        """Transform through intermediates, predict with last step."""
        self._ensure_fitted()
        transformed = self._transform_through(X)
        _, last = self._steps[-1]
        return last.predict(transformed)

    def predict_proba(self, X):
        """Transform through intermediates, predict_proba with last step."""
        self._ensure_fitted()
        _, last = self._steps[-1]
        if not hasattr(last, 'predict_proba'):
            raise ValidationError('Last step does not support predict_proba')
        transformed = self._transform_through(X)
        return last.predict_proba(transformed)

    def score(self, X, y):
        """Transform through intermediates, score with last step."""
        self._ensure_fitted()
        transformed = self._transform_through(X)
        _, last = self._steps[-1]
        return last.score(transformed, y)

    @classmethod
    def load(cls, data):
        """Load a pipeline from a WLRN bundle.

        Each step's artifact is loaded via the global registry.

        Args:
            data: bytes (WLRN bundle)

        Returns:
            Pipeline instance (fitted)
        """
        manifest, toc, blobs = decode_bundle(data)
        steps = []
        for step_info in manifest.get('steps', []):
            name = step_info['name']
            entry = next((t for t in toc if t['id'] == name), None)
            if entry is None:
                raise ValidationError(
                    f'No artifact found for pipeline step "{name}"')
            blob = bytes(blobs[entry['offset']:entry['offset'] + entry['length']])
            estimator = registry_load(blob)
            steps.append((name, estimator))
        pipe = cls(steps)
        pipe._fitted = True
        return pipe

    def save(self):
        """Save pipeline to a WLRN bundle.

        Returns:
            bytes
        """
        self._ensure_fitted()
        manifest = {
            'typeId': PIPELINE_TYPE_ID,
            'steps': [
                {'name': name, 'params': est.get_params()
                 if hasattr(est, 'get_params') else {}}
                for name, est in self._steps
            ],
        }
        artifacts = [
            {'id': name, 'data': est.save(), 'mediaType': 'application/x-wlearn-bundle'}
            for name, est in self._steps
        ]
        return encode_bundle(manifest, artifacts)

    def dispose(self):
        if self._disposed:
            return
        self._disposed = True
        for _, est in self._steps:
            if hasattr(est, 'dispose'):
                est.dispose()

    def get_params(self):
        return {
            name: est.get_params() if hasattr(est, 'get_params') else {}
            for name, est in self._steps
        }

    @property
    def is_fitted(self):
        return self._fitted and not self._disposed

    def _ensure_alive(self):
        if self._disposed:
            raise DisposedError('Pipeline has been disposed.')

    def _ensure_fitted(self):
        self._ensure_alive()
        if not self._fitted:
            raise NotFittedError('Pipeline is not fitted. Call fit() first.')


def _pipeline_loader(manifest, toc, blobs):
    """Registry loader for wlearn.pipeline@1 bundles."""
    steps = []
    for step_info in manifest.get('steps', []):
        name = step_info['name']
        entry = next((t for t in toc if t['id'] == name), None)
        if entry is None:
            raise ValidationError(
                f'No artifact found for pipeline step "{name}"')
        blob = bytes(blobs[entry['offset']:entry['offset'] + entry['length']])
        estimator = registry_load(blob)
        steps.append((name, estimator))
    pipe = Pipeline(steps)
    pipe._fitted = True
    return pipe


register(PIPELINE_TYPE_ID, _pipeline_loader)

File: packages/core/src/pipeline.js
const { Step } = require('./step.js')
const { DisposedError, NotFittedError, ValidationError } = require('./errors.js')
const { encodeBundle, decodeBundle } = require('./bundle.js')
const { register, load: registryLoad } = require('./registry.js')

const PIPELINE_TYPE_ID = 'wlearn.pipeline@1'
let registered = false

/**
 * A pipeline of named estimator steps executed sequentially.
 *
 * Intermediate steps must implement `transform()` (or `fitTransform()`).
 * The last step is the final estimator (predict/score).
 *
 * @example
 * const pipe = new Pipeline([['scaler', scaler], ['clf', model]])
 * pipe.fit(X, y)
 * pipe.predict(X)
 * const bytes = pipe.save()   // WLRN bundle
 * const restored = await load(bytes)
 */
class Pipeline {
  #steps
  #fitted = false
  #disposed = false

  /**
   * @param {Array<[string, Object]>} steps - Array of `[name, estimator]` tuples.
   *   Each estimator must implement the wlearn estimator contract (`fit`, `predict`, `save`, `dispose`).
   * @throws {ValidationError} If steps is empty.
   */
  constructor(steps) {
    this.#steps = steps.map(([name, estimator]) => new Step(name, estimator))
    if (this.#steps.length === 0) {
      throw new ValidationError('Pipeline requires at least one step')
    }
  }

  #ensureAlive() {
    if (this.#disposed) throw new DisposedError('Pipeline has been disposed.')
  }

  #ensureFitted() {
    this.#ensureAlive()
    if (!this.#fitted) throw new NotFittedError('Pipeline is not fitted. Call fit() first.')
  }

  #transformThrough(X) {
    let current = X
    for (let i = 0; i < this.#steps.length - 1; i++) {
      const est = this.#steps[i].estimator
      current = est.transform(current)
    }
    return current
  }

  /**
   * Fit all steps. Intermediate steps are fit-transformed; the last step is fit only.
   * @param {Object} X - Feature matrix (`{ data, rows, cols }` or `number[][]`).
   * @param {Float64Array|Int32Array|number[]} y - Target labels/values.
   * @returns {this}
   */
  fit(X, y) {
    this.#ensureAlive()
    let current = X
    for (let i = 0; i < this.#steps.length - 1; i++) {
      const est = this.#steps[i].estimator
      if (typeof est.fitTransform === 'function') {
        current = est.fitTransform(current, y)
      } else {
        est.fit(current, y)
        current = est.transform(current)
      }
    }
    // Last step: fit only
    const last = this.#steps[this.#steps.length - 1].estimator
    last.fit(current, y)
    this.#fitted = true
    return this
  }

  /**
   * Transform through intermediate steps, then predict with the last step.
   * @param {Object} X - Feature matrix.
   * @returns {Float64Array|Int32Array|Promise<Float64Array|Int32Array>}
   */
  predict(X) {
    this.#ensureFitted()
    const transformed = this.#transformThrough(X)
    return this.#steps[this.#steps.length - 1].estimator.predict(transformed)
  }

  /**
   * Transform through intermediate steps, then call `predictProba` on the last step.
   * @param {Object} X - Feature matrix.
   * @returns {Float64Array|Promise<Float64Array>} Class probability estimates.
   * @throws {ValidationError} If the last step does not support `predictProba`.
   */
  predictProba(X) {
    this.#ensureFitted()
    const last = this.#steps[this.#steps.length - 1].estimator
    if (typeof last.predictProba !== 'function') {
      throw new ValidationError('Last step does not support predictProba')
    }
    const transformed = this.#transformThrough(X)
    return last.predictProba(transformed)
  }

  /**
   * Transform through intermediate steps, then score with the last step.
   * @param {Object} X - Feature matrix.
   * @param {Float64Array|Int32Array|number[]} y - True labels/values.
   * @returns {number|Promise<number>} Score (accuracy for classifiers, R2 for regressors).
   */
  score(X, y) {
    this.#ensureFitted()
    const transformed = this.#transformThrough(X)
    return this.#steps[this.#steps.length - 1].estimator.score(transformed, y)
  }

  /**
   * Serialize the fitted pipeline as a WLRN bundle.
   * Each step's model is saved as a nested artifact.
   * @returns {Uint8Array} Bundle bytes (loadable via `load()` from `@wlearn/core`).
   */
  save() {
    this.#ensureFitted()
    const manifest = {
      typeId: PIPELINE_TYPE_ID,
      steps: this.#steps.map(s => ({
        name: s.name,
        params: s.estimator.getParams()
      }))
    }
    const artifacts = this.#steps.map(s => ({
      id: s.name,
      data: s.estimator.save(),
      mediaType: 'application/x-wlearn-bundle'
    }))
    return encodeBundle(manifest, artifacts)
  }

  /**
   * Load a pipeline from WLRN bundle bytes.
   * Dispatches each step's blob through the global registry to reconstruct estimators.
   * @param {Uint8Array} bytes - Bundle bytes produced by `pipeline.save()`.
   * @returns {Promise<Pipeline>} A fitted pipeline ready for predict/score.
   */
  static async load(bytes) {
    const { manifest, toc, blobs } = decodeBundle(bytes)
    const steps = []
    for (const stepInfo of manifest.steps) {
      const tocEntry = toc.find(t => t.id === stepInfo.name)
      if (!tocEntry) {
        throw new ValidationError(`No artifact found for pipeline step "${stepInfo.name}"`)
      }
      const blob = blobs.subarray(tocEntry.offset, tocEntry.offset + tocEntry.length)
      const estimator = await registryLoad(blob)
      steps.push([stepInfo.name, estimator])
    }
    const pipeline = new Pipeline(steps)
    pipeline.#fitted = true
    return pipeline
  }

  /** Dispose all step estimators and mark the pipeline as disposed. */
  dispose() {
    if (this.#disposed) return
    this.#disposed = true
    for (const step of this.#steps) {
      step.estimator.dispose()
    }
  }

  getParams() {
    const params = {}
    for (const step of this.#steps) {
      params[step.name] = step.estimator.getParams()
    }
    return params
  }

  setParams(p) {
    this.#ensureAlive()
    for (const step of this.#steps) {
      if (p[step.name]) {
        step.estimator.setParams(p[step.name])
      }
    }
    return this
  }

  get capabilities() {
    return this.#steps[this.#steps.length - 1].estimator.capabilities
  }

  get isFitted() { return this.#fitted }

  static registerLoader() {
    if (registered) return
    registered = true
    register(PIPELINE_TYPE_ID, (manifest, toc, blobs) => {
      return Pipeline._loadFromParts(manifest, toc, blobs)
    })
  }

  static async _loadFromParts(manifest, toc, blobs) {
    const steps = []
    for (const stepInfo of manifest.steps) {
      const tocEntry = toc.find(t => t.id === stepInfo.name)
      if (!tocEntry) {
        throw new ValidationError(`No artifact found for pipeline step "${stepInfo.name}"`)
      }
      const blob = blobs.subarray(tocEntry.offset, tocEntry.offset + tocEntry.length)
      const estimator = await registryLoad(blob)
      steps.push([stepInfo.name, estimator])
    }
    const pipeline = new Pipeline(steps)
    pipeline.#fitted = true
    return pipeline
  }
}

// Auto-register pipeline loader
Pipeline.registerLoader()

module.exports = { Pipeline }
